persistent_stream.go

```diff
@@ -0,0 +1,334 @@
+	callID, _ := uuid.FromBytes(req.CallId)
+	rc := make(chan *pb.PersistentConnectionResponse)
+
+	go func() {
+		defer close(rc)
+
+		if err := ggio.NewDelimitedWriter(s).WriteMsg(req); ctx.Err() != nil {
+			return
+		} else if err != nil {
+			rc <- errorUnaryCall(callID, err)
+			return
+		}
+
+		remoteResp := &pb.PersistentConnectionRequest{}
+		if err := ggio.NewDelimitedReader(s, network.MessageSizeMax).ReadMsg(remoteResp); ctx.Err() != nil {
+			return
+		} else if err != nil {
+			rc <- errorUnaryCall(callID, err)
+			return
+		}
+
+		resp := okUnaryCallResponse(callID)
+		resp.Message = &pb.PersistentConnectionResponse_CallUnaryResponse{
+			CallUnaryResponse: remoteResp.GetUnaryResponse(),
+		}
+
+		select {
+		case rc <- resp:
+			return
+
+		case <-ctx.Done():
+			return
+		}
+	}()
+
+	return rc
+}
+
+// awaitReadFail writers to a semaphor channel if the given io.Reader fails to
+// read before the context was cancelled
+func awaitReadFail(ctx context.Context, r io.Reader) <-chan struct{} {
+	semaphor := make(chan struct{})
+
+	go func() {
+		defer close(semaphor)
+
+		buff := make([]byte, 1)
+		if _, err := r.Read(buff); err != nil {
+			select {
+			case semaphor <- struct{}{}:
+			case <-ctx.Done():
+			}
+		}
+	}()
+
+	return semaphor
+}
+
+// getPersistentStreamHandler returns a lib-p2p stream handler tied to a
+// given persistent client stream
+func (d *Daemon) getPersistentStreamHandler(cw ggio.Writer) network.StreamHandler {
+	return func(s network.Stream) {
+		defer s.Close()
+
+		req := &pb.PersistentConnectionRequest{}
+		if err := ggio.NewDelimitedReader(s, network.MessageSizeMax).ReadMsg(req); err != nil {
+			log.Debugw("failed to read proto from incoming p2p stream", "error", err)
+			return
+		}
+
+		// now the peer field stores the caller's peer id
+		req.GetCallUnary().Peer = []byte(s.Conn().RemotePeer())
+
+		callID, err := uuid.FromBytes(req.CallId)
+		if err != nil {
+			log.Debugw("bad call id in p2p handler", "error", err)
+			return
+		}
+
+		rc := make(chan *pb.PersistentConnectionRequest)
+		d.responseWaiters.Store(callID, rc)
+		defer d.responseWaiters.Delete(callID)
+
+		ctx, cancel := context.WithCancel(d.ctx)
+		defer cancel()
+
+		resp := &pb.PersistentConnectionResponse{
+			CallId: req.CallId,
+			Message: &pb.PersistentConnectionResponse_RequestHandling{
+				RequestHandling: req.GetCallUnary(),
+			},
+		}
+		if err := cw.WriteMsg(resp); err != nil {
+			log.Debugw("failed to write message to client", "error", err)
+			return
+		}
+
+		select {
+		case <-awaitReadFail(ctx, s):
+			if err := cw.WriteMsg(
+				&pb.PersistentConnectionResponse{
+					CallId: callID[:],
+					Message: &pb.PersistentConnectionResponse_Cancel{
+						Cancel: &pb.Cancel{},
+					},
+				},
+			); err != nil {
+				log.Debugw("failed to write to client", "error", err)
+				return
+			}
+			return
```

Comment:
```suggestion
			}
```

This can be simplified.